File: Visual Basic/Database/SummaryByQuarter.vb
Imports Microsoft.VisualBasic
Imports System
Imports System.Data
Namespace Aspose.Cells.Demos
	''' <summary>
	''' Summary description for SummaryByQuarter.
	''' </summary>
	Public Class SummaryByQuarter
		Inherits DbBase
		Public Sub New(ByVal path As String)
			MyBase.New(path)

		End Sub

		Public Function CreateSummaryByQuarter() As Workbook
			Try
				DBInit()
			Catch
			Finally
				If Me.oleDbConnection1 IsNot Nothing Then
					Me.oleDbConnection1.Close()
				End If
			End Try

			'Open the template file
			Dim designerFile As String = MapPath("~/Designer/Northwind.xls")
			Dim workbook As New Workbook(designerFile)

			Try
				'Specify SQL and execute the query to fill the datatable
				Me.oleDbSelectCommand1.CommandText = "SELECT COUNT(Orders.OrderID) AS Orders, " & ControlChars.CrLf & "					SUM([Order Subtotals].Subtotal) AS Sales, FORMAT(Orders.ShippedDate, 'yyyy/Q') AS Quarter " & ControlChars.CrLf & "					FROM Orders INNER JOIN [Order Subtotals] ON Orders.OrderID = [Order Subtotals].OrderID " & ControlChars.CrLf & "					WHERE (Orders.ShippedDate IS NOT NULL) GROUP BY FORMAT(Orders.ShippedDate, 'yyyy/Q')"
				Me.oleDbDataAdapter1.Fill(Me.dataTable1)
			Catch
			Finally
				If Me.oleDbConnection1 IsNot Nothing Then
					Me.oleDbConnection1.Close()
				End If
			End Try
			'Get the worksheet
			Dim sheet As Worksheet = workbook.Worksheets("Sheet13")
			'Name the sheet
			sheet.Name = "Summary By Quarter"
			'Get the cells
			Dim cells As Cells = sheet.Cells

			'Create an arry of datatable with fields
			Dim quarterSummary(3) As DataTable
			For i As Integer = 0 To 3
				quarterSummary(i) = New DataTable()
				quarterSummary(i).Columns.Add("YearOrQuarter", GetType(Integer))
				quarterSummary(i).Columns.Add("Orders", GetType(Integer))
				quarterSummary(i).Columns.Add("Sales", GetType(Decimal))
			Next i

			'Adding some records to the datatables
			For i As Integer = 0 To Me.dataTable1.Rows.Count - 1
				Dim strQuarter As String = CStr(Me.dataTable1.Rows(i)("Quarter"))
				Dim quarter As Integer = Integer.Parse(strQuarter.Substring(strQuarter.Length - 1))
				Dim row As DataRow = quarterSummary(quarter - 1).NewRow()
				row("YearOrQuarter") = Integer.Parse(strQuarter.Substring(0, 4))
				row("Sales") = Me.dataTable1.Rows(i)("Sales")
				row("Orders") = Me.dataTable1.Rows(i)("Orders")
				quarterSummary(quarter - 1).Rows.Add(row)
			Next i

			'Replace some values in the workbook
			For i As Integer = 0 To 3
				workbook.Replace("&summary" & (i + 1).ToString(), quarterSummary(i))
			Next i
			'Remove the unnecessary worksheets in the workbook
            Dim iworkbook As Integer = 0
            Do While iworkbook < workbook.Worksheets.Count
                sheet = workbook.Worksheets(iworkbook)
                If sheet.Name <> "Summary By Quarter" Then
                    workbook.Worksheets.RemoveAt(iworkbook)
                    iworkbook -= 1
                End If
                iworkbook += 1
            Loop
			'Get the generated workbook
			Return workbook
		End Function

	End Class
End Namespace

File: C Sharp/Database/SummaryByQuarter.cs
using System;
using System.Data;
namespace Aspose.Cells.Demos
{
    /// <summary>
    /// Summary description for SummaryByQuarter.
    /// </summary>
    public class SummaryByQuarter : DbBase
    {
        public SummaryByQuarter(string path)
            : base(path)
        {

        }

        public Workbook CreateSummaryByQuarter()
        {
            try
            {
                DBInit();
            }
            catch
            {
            }
            finally
            {
                if (this.oleDbConnection1 != null)
                    this.oleDbConnection1.Close();
            }

            //Open the template file
		    string designerFile = MapPath("~/Designer/Northwind.xls");						
            Workbook workbook = new Workbook(designerFile);

            try
            {
                //Specify SQL and execute the query to fill the datatable
                this.oleDbSelectCommand1.CommandText = @"SELECT COUNT(Orders.OrderID) AS Orders, 
					SUM([Order Subtotals].Subtotal) AS Sales, FORMAT(Orders.ShippedDate, 'yyyy/Q') AS Quarter 
					FROM Orders INNER JOIN [Order Subtotals] ON Orders.OrderID = [Order Subtotals].OrderID 
					WHERE (Orders.ShippedDate IS NOT NULL) GROUP BY FORMAT(Orders.ShippedDate, 'yyyy/Q')";
                this.oleDbDataAdapter1.Fill(this.dataTable1);
            }
            catch
            {
            }
            finally
            {
                if (this.oleDbConnection1 != null)
                    this.oleDbConnection1.Close();
            }
            //Get the worksheet
            Worksheet sheet = workbook.Worksheets["Sheet13"];
            //Name the sheet
            sheet.Name = "Summary By Quarter";
            //Get the cells
            Cells cells = sheet.Cells;

            //Create an arry of datatable with fields
            DataTable[] quarterSummary = new DataTable[4];
            for (int i = 0; i < 4; i++)
            {
                quarterSummary[i] = new DataTable();
                quarterSummary[i].Columns.Add("YearOrQuarter", typeof(int));
                quarterSummary[i].Columns.Add("Orders", typeof(int));
                quarterSummary[i].Columns.Add("Sales", typeof(decimal));
            }

            //Adding some records to the datatables
            for (int i = 0; i < this.dataTable1.Rows.Count; i++)
            {
                string strQuarter = (string)this.dataTable1.Rows[i]["Quarter"];
                int quarter = int.Parse(strQuarter.Substring(strQuarter.Length - 1));
                DataRow row = quarterSummary[quarter - 1].NewRow();
                row["YearOrQuarter"] = int.Parse(strQuarter.Substring(0, 4));
                row["Sales"] = this.dataTable1.Rows[i]["Sales"];
                row["Orders"] = this.dataTable1.Rows[i]["Orders"];
                quarterSummary[quarter - 1].Rows.Add(row);
            }

            //Replace some values in the workbook
            for (int i = 0; i < 4; i++)
            {
                workbook.Replace("&summary" + (i + 1).ToString(), quarterSummary[i]);
            }
            //Remove the unnecessary worksheets in the workbook
            for (int i = 0; i < workbook.Worksheets.Count; i++)
            {
                sheet = workbook.Worksheets[i];
                if (sheet.Name != "Summary By Quarter")
                {
                    workbook.Worksheets.RemoveAt(i);
                    i--;
                }
            }
            //Get the generated workbook
            return workbook;
        }

    }
}


